### ISSUE TITLE
Build a theme with BackboneJS, UnderscoreJS and REST backend.

### ISSUE THREAD

**Issue by chris001** [salesagility/SuiteCRM#3183]
Hundreds of bugs/issues here, burning up many thousands of hours of support and user community, and development time, are ultimately come down to a having inherited SugarCRM 6.5's 2004-era web user interface architecture.  Web user interfaces have improved dramatically in the last 13 years, and SuiteCRM needs to keep up with user expectations and update.

#### Issue

It appears that you can't rip out the web user interface and replace it with a newer one, without breaking the functionality of add-on modules, and the core itself.

#### Expected Behavior

The models, views, and controllers should be separated out from one another, and work with any dependencies on how the others work internally.

#### Possible Fix

* BackboneJS is a notable javascript framework, about 1000 lines works with jquery. Provides enough structure to cleanly separate out the models, views, and controllers.  Events and listeners.  Used by SugarCRM 7.x and thousands of other web apps.  Restful routes, crud operations, etc.
* UnderscoreJS is a great HTML templating engine - also used by SugarCRM 7.x, and thousands of more projects.

The end result would be a much more reliable, and easier to develop for, single page application, much faster to debug, and probably attract 10X more users and module developers.  Suite is now almost a single page application, thanks to the AJAX.

More information about BackboneJS :
https://www.youtube.com/watch?v=UAl_N62gKmM

**Comment by pribeiro42** [comment]
Isn't this the goal for the future 8.0 version ?

**Comment by chris001** [comment]
@pribeiro42   The roadmap only says, New Rest interface, New frontend.  Missing details on what those will be, how it is developed, when, by who.

**Comment by pribeiro42** [comment]
@chris001 I see... so, you're proposing a concrete implementation of the Roadmap intentions 👍

**Comment by pgorod** [comment]
I worry that in SuiteCRM, a Theme isn't really just a Theme. It's not something you can use to just change the aesthetics of the app. It's too tightly tied to the logic and structure underneath, and so we never really get more than one theme at a given time. 

Notice that no 3rd party ever contributes themes, and SalesAgility only really support one theme at any given time (it was SuiteR, now it's SuiteP). No wonder - it's not just a matter of changing colors and fonts, a Theme for us is a big, big project, with months of debugging and fine-tuning needed. See how many things stopped working with the new Theme, we're still fixing them. 

I'm sticking with SuiteR for functional reasons (compactness), and some aesthetic preference (with my customizations), but I worry that SuiteP has technological improvements that I'm not taking advantage of.

I liked how in the old SugarCRM we could easily change it's color palette from the top menu, you could easily change colors every day if you felt like it.

It seems like, architecturally speaking, this project took a wrong turn somewhere in terms of Themes and front-end...

**Comment by pribeiro42** [comment]
@pgorod Not only Themes and front-end, but, like @chris001 says, the back-end should be a lot more modular than currently is.. In fact, i think one of the worst problems SuiteCRM has is that there is not really a difference between back-end and front-end.. We should have a document to define the roles and scopes of each part...

Of course, architecture changes are allways disruptive and demand a lot of work, both from SalesAgility and 3rd party module developers, but I, for one, am willing to pay the price if that means a better and improved product for everyone...

**Comment by chris001** [comment]
@pribeiro42 @pgorod 
If you watch the youtube video posted above about backbone and underscore, and look at the overall architecture of Sugar 6.5 and Sugar 7.x, you can see that the intended way in 6.5 was to separate out the view piece from the data model and business logic pieces.  The great thing about Sugar 7.x they have a new rule, all add-on modules, and even the stock web user interface, must communicate with the core data model/business logic, thru the REST interface.  No longer are you allowed to do raw SQL calls on the data, because you don't get direct access to it in that way.  This is how SuiteCRM 8 should be.  This is the type of application architecture which lets you go forward in an organized powerful manner, and not have things brittle and breaking everywhere, just because some 3rd party modules are depending on the internal way the database was structured at one time. which it shouldn't have, and when you updated that area of the database, those modules all break.

**Comment by samus-aran** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Build a theme with BackboneJS, UnderscoreJS and REST backend.](https://trello.com/c/W2LbBjpb/22-build-a-theme-with-backbonejs-underscorejs-and-rest-backend)

**Comment by samus-aran** [comment]
We have now set up a new home for suggestions at Trello.  All github issues that were labeled 'suggestion' have been moved and will be closed.  Certain ones will be progressed within the new Suggestion Box and may be re-opened.

Announcement of moving Suggestions:
https://suitecrm.com/forum/suggestion-box/13691-moving-suggestions-from-github-to

New SuiteCRM Suggestion Box
https://trello.com/b/Ht7LbMqw/suitecrm-suggestion-box
